src/thunderbots/software/ai/hl/stp/play/play.h

```diff
@@ -0,0 +1,76 @@
+
+/**
+ * The PlayFactory is an Abstract class that provides an interface for Play Factories
+ * to follow. This makes it easy to maintain a list of factories and get the corresponding
+ * plays through the generic interface.
+ */
+class PlayFactory
+{
+   public:
+    virtual std::shared_ptr<Play> getInstance() = 0;
+};
+
+/**
+ * In the STP framework, a Play is essentially a collection of tactics that represent some
+ * "team-wide" goal. It can be thought of like a tradition play in soccer, such as an
+ * offensive play, defensive play, or a specific play used for corner kicks.
+ *
+ * When we are running autonomously, different Plays are selected at different times
+ * based on the state of the world, and this switching of Plays is what allows our AI to
+ * play a full game of soccer, as long a we provide a Play for any given scenario.
+ */
+class Play
+{
+   public:
+    /**
+     * Returns whether or not this Play can be started.
+     *
+     * @param world The current state of the world
+     * @return true if this Play can be started, and false otherwise
+     */
+    virtual bool isApplicable(const World& world) = 0;
+
+    virtual bool invariantHolds(const World& world) = 0;
+    virtual bool hasFailed(const World& world)      = 0;
+
+    virtual std::vector<std::unique_ptr<Tactic>> getTactics(const World& world) = 0;
+
+    virtual std::string name() = 0;
+
+    static std::vector<std::string> getPlayNames();
+
+    // Factory stuff
+    static const std::vector<std::shared_ptr<PlayFactory>>& getRegistry();
+    static void registerPlay(std::shared_ptr<PlayFactory> play_factory);
+
+   private:
+    static std::vector<std::shared_ptr<PlayFactory>>& getMutableRegistry();
+};
+
+#define REGISTER_PLAY(klass)                                                             \
+    class klass##Factory : public PlayFactory                                            \
+    {                                                                                    \
+       public:                                                                           \
+        klass##Factory()                                                                 \
+        {                                                                                \
+            Play::registerPlay(std::shared_ptr<klass##Factory>(this));                   \
+        }                                                                                \
+        std::shared_ptr<Play> getInstance() override                                     \
+        {                                                                                \
+            return std::make_shared<klass>();                                            \
+        }                                                                                \
+    };                                                                                   \
+    static klass##Factory global_##klass##Factory;
```

Comment:
This seems like a bit of a hack, and also like a textbook case for templating, but if you'd  rather not get into that, we should at least change `klass` to something more descriptive, such as `play_class_name` (unless `klass` means something special?).

Reply:
Does not mean something special, will change. Templating still won't entirely replace the auto-registration because the static initialization is still required. Unless you mean just replacing the implementation of the macro?

Reply:
ie. something like (just off the top of my head)
```
template <typename PLAY_T>
class PlayFactory {
    public :
        std::shared_ptr<PlayFactory> getFactoryInstance()
        {
            if (!factory_instance){
                factory_instance = std::make_shared<PlayFactory<PLAY_T>>;
                Play::registerPlay(std::shared_ptr<PlayFactory<PLAY_T>>(
                            *factory_instance)); 
            }
            return factory_instance
        }

        std::shared_ptr<Play> getInstance() 
        {
            return std::make_shared<PLAY_T>();
        }

    private:
        static std::shared_ptr<PlayFactory<PLAY_T>> factory_instance;

        // Default constructor is private to force caller to use 
        // getFactoryInstance
        PlayFactory();

}
```
and get rid of this macro entirely.

Reply:
I took a quick crack at this and it might be a bit trickier that I thought, mostly due to nitty-gritty declaration ordering stuff. In the interest of moving forward for now I'll update the `klass` name to something more descriptive. I have created a ticket to investigate templating in the future #35